[BUG] Fixed php8 error when calling hasDirect whit non-object or non-…

had1z opened the pull request:
Fixes #23

mpetrovich commented:
Thanks for making the fix!

Were you able to run the full set of unit tests in PHP 8? (via `make test`)

had1z commented:
> Thanks for making the fix!
> 
> Were you able to run the full set of unit tests in PHP 8? (via `make test`)

The test suit has some problems with php 7.3 and higher. You can see that in travis job which ran against my PR. But it's not related to this issue.
However I think my commit has a problem, because I didn't check `method_exists` if the argument is a class name string.

had1z commented:
This is the problem that we have in php 8.0.5. I think it could be addressed in another PR.
```
PHP Fatal error:  Uncaught Error: Call to undefined function each() in /home/hadi/Projects/Dash/vendor/phpunit/phpunit/src/Util/Getopt.php:38
```

had1z commented:
> However I think my commit has a problem, because I didn't check `method_exists` if the argument is a class name string.
I fixed this issue in another commit.

had1z commented:
> > However I think my commit has a problem, because I didn't check `method_exists` if the argument is a class name string.
> 
> I fixed this issue in another commit.

Ok, it has another issue :)
I didn't check the `$input` to be `string` before putting it in `class_exists`.

had1z commented:
> > > However I think my commit has a problem, because I didn't check `method_exists` if the argument is a class name string.
> > 
> > 
> > I fixed this issue in another commit.
> 
> Ok, it has another issue :)
> I didn't check the `$input` to be `string` before putting it in `class_exists`.

Fixed this issue new commit. 
I tested this last commit against php version up to 7.4, version 8 has an issue related to PHPUnit.

mpetrovich commented:
PHPUnit will probably need to be upgraded to a version that works with PHP 8. If true, we'll need to make Dash require PHP 8+ going forward. I can handle that version constraint.

If you're comfortable upgrading PHPUnit, try doing that and re-running in PHP 8. If not, let me know.

had1z commented:
> PHPUnit will probably need to be upgraded to a version that works with PHP 8. If true, we'll need to make Dash require PHP 8+ going forward. I can handle that version constraint.
> 
> If you're comfortable upgrading PHPUnit, try doing that and re-running in PHP 8. If not, let me know.

Hi Mike.
I tried to do this but PHPUnit's newer version do not use `PHPUnit_Framework_TestCase` anymore. So all test suites are going to need change to use `PHPUnit\Framework\TestCase` instead.
I did a test with `PHPUnit\Framework\TestCase` and there was too many errors in test classes, the test classes should be refactored.

had1z commented:
I found the problem and it wasn't related to test suites. It was the same problem which I fixed in this PR.
I found the same issue in `getDirect`.
But there is still too many *warnings* in running tests, so the test suites should be refactored.

mpetrovich commented:
I was able to incorporate your changes into [v4.0.0](https://github.com/mpetrovich/dash/releases/tag/v4.0.0)! Thank you for your help :)

mpetrovich commented:
Argh, for some reason my git wrangling clobbered your original commits, so you won't get the proper attribution :(
